### ISSUE TITLE
Proper shutdown sequence?

### ISSUE THREAD

**Issue by auras** [4thline/cling#170]
What is the way to shutdown the service without errors?

Right now I'm doing it like this:

```
        if (mUpnpService != null) {
            mUpnpService.getRegistry().removeListener(mRegistryListener);
            mUpnpService.getRegistry().removeAllLocalDevices();
            mUpnpService.getRegistry().removeAllRemoteDevices();
        }

        synchronized (mSubscriptionCallbackList) {
            for (SubscriptionCallback callback : mSubscriptionCallbackList) {
                callback.end();
            }

            mSubscriptionCallbackList.clear();
        }

        try {
            getContext().unbindService(this);
        }
        catch (IllegalArgumentException ignore) { } // Already unbound
        mUpnpService = null;
```

But then I get a lot of messages like this:
`I/pnpServiceConfiguration: Thread pool rejected execution of class org.eclipse.jetty.io.nio.SelectorManager$SelectSet$1
`

Which leads me to believe that it's not shutting down the sockets, the services.

I also remove all references to Device objects so I can't control them anymore and do any actions.

P.S.: I'm using AndroidUpnpServiceImpl as is, not extending it.

**Comment by christianbauer** [comment]
The shutdown() method:

http://4thline.org/projects/cling/core/manual/cling-core-manual.html#chapter.GettingStarted

**Comment by auras** [comment]
I've figured out why it's not shutting down:

AndroidRouter.disable() cannot acquire a writeLock and thus it's not shutting down properly. I'm trying to investigate why. Any reason why acquiring a writeLock would fail, but not for a readLock?

**Comment by auras** [comment]
I'm guessing it can happen because the timeout for StreamClient is a lot larger (60s) vs the lock timeout (6-15s). 
Am I right?

**Comment by auras** [comment]
Yeah, so there are scenarios where it's trying to communicate with something, so it has acquired a lock, but the timeout for send() is larger than the timeout for lock() so disable() will fail and the service will remain in limbo. It can't be enabled nor disabled, nor can you catch if shutdown() did work or not.

So my dirty solution is to kill all the cling threads so that the send commands are interrupted and the locks are released:

```
        new Thread() {
            public void run() {
                mUpnpService = null;

                Set<Thread> threadSet = Thread.getAllStackTraces().keySet();

                for (Thread thread : threadSet) {
                    if (thread.getName().startsWith("cling")) {
                        Log.d(DEBUG_TAG, "Killing thread #" + thread.getId() + " " + thread.getName());
                        thread.interrupt();
                    }
                }

                try {
                    getContext().unbindService(UpnpApp.this);
                }
                catch (IllegalArgumentException ignore) { } // Already unbound
            }
        }.start();
```

**Comment by auras** [comment]
Of course, the other solution is to increase the lock timeout to something larger than 60s and wait for the send timeout to kick in.

Another better solution would be to have the streamClient run in a separate thread and stop it.

**Comment by christianbauer** [comment]
I haven't looked at this problem in a while but there have been many (somewhat different) reports about shutdown problems in the past and patches. Another one can't hurt. I have noticed that the major Cling user BubbleUPnP has problems now on my Nexus 6P, restarting the app doesn't work properly.
